File: python/finance.py

import time
from functools import wraps

def timing(method):
    """Decorator for timing functions
    Usage:
    @timing
    def function(a):
        pass
    """

    @wraps(method)
    def wrapper(*args, **kwargs):
        start = time.time()
        result = method(*args, **kwargs)
        end = time.time()
        print('function:%r took: %2.22f secs' % (method.__name__,  end - start))
        return result
    return wrapper


@timing
def selectStock(savings: int, current_value: list, future_value: list) -> int:
    """Method to find the stocks to buy to maximise the profit.
    The loop is going to sum every each element with each of the rest others.
    If two elements minus saving is not negative or not sum of the same number, save it."""
    result = dict()
    for index, el in enumerate(current_value):
        for index_2, el_2 in enumerate(current_value):
            sum = el + el_2
            if savings - sum >= 0 and index != index_2:
                result[sum] = [index, index_2]
    return check_profit(result, future_value)

def check_profit(current: dict, future_value: list) -> int:
    """current: {242: [2, 1], 230: [4, 1], 206: [4, 2]}
    Add the elements from future_value in the same postion as the current
    and substract them to find out the most profitable stock."""
    result = []
    for el, index in current.items():
        future_profit = future_value[index[0]] + future_value[index[1]]
        result.append(future_profit - el)
    return max(result)

@timing
def select_stock_comprehension(savings: int, current_value: list, future_value: list) -> int:
    """Same method than selectStock() but we are using list comprehension.
    Slightly slower than selectStock() on average 0.000001 seconds."""
    result = {
        el + el_2:[index, index_2]
        for index, el in enumerate(current_value)
        for index_2, el_2 in enumerate(current_value)
        if savings - (el + el_2) >= 0 and index != index_2
    }
    return check_profit_comprehension(result, future_value)


def check_profit_comprehension(current: dict, future_value: list) -> int:
    """Same method as check_profit() but we are using list comprehension.
    Slightly slower than selectStock() on average 0.000001 seconds."""
    return max([
        (future_value[index[0]] + future_value[index[1]]) - el
        for el, index in current.items()
    ])

File: typescript/src/finance.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.selectStock = void 0;
var startTime = performance.now();
function selectStock(savings, currentValue, futureValue) {
    let result = {};
    let sum;
    for (const [index, val] of Object.entries(currentValue)) {
        for (const [index_2, val_2] of Object.entries(currentValue)) {
            sum = val + val_2;
            if (savings - sum >= 0 && index != index_2) {
                result[sum] = [index, index_2];
            }
        }
    }
    return checkProfit(result, futureValue);
}
exports.selectStock = selectStock;
function checkProfit(current = {}, futureValue) {
    let result = [];
    for (var key in current) {
        let index = current[key][0];
        let index2 = current[key][1];
        let futureProfit = futureValue.slice(index)[0] + futureValue.slice(index2)[0];
        result.push(futureProfit - (+key)); //The right-hand side of an arithmetic operation must be of type 'any', 'number', 'bigint' or an enum type.
    }
    return result.reduce((a, b) => Math.max(a, b));
    // Object.entries(current).forEach((key, value) => {
    // let index:any = value[0];
    // let index2:any = value[1];
    // console.log("index--->", value, "index2--> ", value, "key: ", key)
    // let futureProfit:number = futureValue.slice(index)[0] + futureValue.slice(index2)[0];
    // console.log(futureValue.slice(index), " => ", futureValue.slice(index2));
    // result.push(futureProfit - value)
    // });
    // for (const [index, val] of Object.entries(current)) {
    //     let futureProfit:number = futureValue[index[0]] + futureValue[index[1]]
    //     result.push(futureProfit - val)
    // }
    // return Math.max.apply(result)
}
var endTime = performance.now();
console.log(`Call to doSomething took ${endTime - startTime} milliseconds`);
